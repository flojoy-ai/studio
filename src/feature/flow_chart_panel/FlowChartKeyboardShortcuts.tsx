import ReactFlow, {
  ConnectionLineType,
  Edge,
  NodeTypes,
  OnNodesDelete,
  Node,
  useReactFlow,
  EdgeTypes,
  OnInit,
  OnNodesChange,
  OnEdgesChange,
  OnConnect,
  NodeDragHandler,
} from "reactflow";
import useKeyboardShortcut from "@src/hooks/useKeyboardShortcut";
import { AddNodeBtn } from "@src/AddNodeBtn";
import { ClearCanvasBtn } from "./components/clear-canvas-btn/ClearCanvasBtn";
import { useFlowChartGraph } from "@src/hooks/useFlowChartGraph";
import { useFlowChartState } from "@src/hooks/useFlowChartState";
import { ElementsData } from "@src/feature/flow_chart_panel/types/CustomNodeProps";
import "reactflow/dist/style.css";

type FlowChartKeyboardShortcutsProps = {
  nodes: Node<ElementsData>[];
  nodeTypes: NodeTypes;
  edges: Edge[];
  edgeTypes: EdgeTypes;
  onInit: OnInit;
  onNodesChange: OnNodesChange;
  onEdgesChange: OnEdgesChange;
  onConnect: OnConnect;
  handleNodeDrag: NodeDragHandler;
  handleNodesDelete: OnNodesDelete;
};

const FlowChartKeyboardShortcuts = ({
  nodes,
  nodeTypes,
  edges,
  edgeTypes,
  onInit,
  onNodesChange,
  onEdgesChange,
  onConnect,
  handleNodeDrag,
  handleNodesDelete,
}: FlowChartKeyboardShortcutsProps) => {
  const { zoomIn, zoomOut, fitView } = useReactFlow();
  useKeyboardShortcut("ctrl", "=", zoomIn);
  useKeyboardShortcut("ctrl", "-", zoomOut);
  useKeyboardShortcut("ctrl", "1", fitView)

  const { setIsSidebarOpen } = useFlowChartState();
  const { setNodes, setEdges } = useFlowChartGraph();

  return (
    <ReactFlow
      style={{
        position: "fixed",
        height: "100%",
        width: "50%",
      }}
      nodes={nodes}
      nodeTypes={nodeTypes}
      edges={edges}
      edgeTypes={edgeTypes}
      connectionLineType={ConnectionLineType.Step}
      onInit={onInit}
      onNodesChange={onNodesChange}
      onEdgesChange={onEdgesChange}
      onConnect={onConnect}
      onNodeDragStop={handleNodeDrag}
      onNodesDelete={handleNodesDelete}
    >
      <div
        className="top-row"
        style={{
          display: "flex",
          justifyContent: "space-between",
          zIndex: 100,
        }}
      >
        <AddNodeBtn setIsSidebarOpen={setIsSidebarOpen} />
        <ClearCanvasBtn setNodes={setNodes} setEdges={setEdges} />
      </div>
    </ReactFlow>
  );
};

export default FlowChartKeyboardShortcuts;
