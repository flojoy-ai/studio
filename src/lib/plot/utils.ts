export function clamp(x: number, lo: number, hi: number) {
  return Math.min(Math.max(x, lo), hi);
}
