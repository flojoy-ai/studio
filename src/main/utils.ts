import * as http from "http";
import * as fs from "fs";
import { isIP } from "net";
import { execCommand } from "./executor";
import { Command } from "./command";
import { app, dialog } from "electron";
import { join } from "path";
import { killCaptain } from "./python";
import log from "electron-log";
import { ChildProcess } from "node:child_process";
import { Err, Result, Ok } from "../types/result";

export const isPortFree = (port: number) =>
  new Promise((resolve) => {
    const server = http
      .createServer()
      .listen(port, "127.0.0.1", () => {
        server.close();
        resolve(true);
      })
      .on("error", () => {
        resolve(false);
      });
  });

export const killProcess = async (port: number) => {
  await execCommand(
    new Command({
      darwin: `kill -9 $(lsof -t -i :${port})`,
      linux: `kill -9 $(lsof -t -i :${port})`,
      win32: `FOR /F "tokens=5" %i IN ('netstat -aon ^| find "${port}"') DO Taskkill /F /PID %i`,
    }),
  );
};

export const ping = async (addr: string) => {
  if (isIP(addr) === 0) {
    throw new Error(`Invalid IP address: ${addr}`);
  }

  return await execCommand(
    new Command({
      win32: `ping -n 1 ${addr}`,
      darwin: `ping -c 1 ${addr}`,
      linux: `ping -c 1 ${addr}`,
    }),
    { quiet: true },
  );
};

export const netstat = async () => {
  return await execCommand(
    new Command({
      win32: "netstat",
      darwin: "netstat",
      linux: "netstat",
    }),
    { quiet: true },
  );
};

export const ifconfig = async () => {
  return await execCommand(
    new Command({
      win32: "ipconfig /all",
      darwin: "ifconfig",
      linux: "ifconfig",
    }),
    { quiet: true },
  );
};

export const writeFileSync = (_, filePath: string, text: string): void => {
  fs.writeFileSync(filePath, text);
};

export const pickDirectory = async (): Promise<string> => {
  const handler = await dialog.showOpenDialog({
    properties: ["openDirectory"],
  });
  return handler.canceled ? "" : handler.filePaths[0];
};

export const getCustomBlocksDir = async () => {
  const filePath = join(app.getPath("home"), ".flojoy/custom_blocks_path.txt");

  if (!fs.existsSync(filePath)) {
    return undefined;
  }

  const blocksPath = fs
    .readFileSync(filePath, { encoding: "utf-8" })
    .toString();

  if (!fs.existsSync(blocksPath)) {
    return undefined;
  }

  return blocksPath;
};

export const cacheCustomBlocksDir = (_, dirPath: string) => {
  const flojoyDir = join(app.getPath("home"), ".flojoy");
  if (!fs.existsSync(flojoyDir)) {
    fs.mkdirSync(flojoyDir);
  }
  const cacheFilePath = join(flojoyDir, "custom_blocks_path.txt");
  fs.writeFileSync(cacheFilePath, dirPath);
};

export const openTestPicker = (): Promise<
  { filePath: string; fileContent: string } | undefined
> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: ["json", "py"],
            name: "Test",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const openFilePicker = (
  _,
  allowedExtensions: string[] = ["json"],
): Promise<{ filePath: string; fileContent: string } | undefined> => {
  return new Promise((resolve, reject) => {
    try {
      const selectedPaths = dialog.showOpenDialogSync(global.mainWindow, {
        properties: ["openFile"],
        filters: [
          {
            extensions: allowedExtensions,
            name: "allowed extensions",
          },
        ],
      });
      if (selectedPaths && selectedPaths?.length > 0) {
        const fileContent = fs.readFileSync(selectedPaths[0], {
          encoding: "utf-8",
        });
        resolve({
          filePath: selectedPaths[0],
          fileContent,
        });
        resolve(undefined);
      }
    } catch (error) {
      reject(String(error));
    }
  });
};

export const cleanup = async () => {
  const captainProcess = global.captainProcess as ChildProcess;
  log.info(
    "Cleanup function invoked, is captain running? ",
    !(captainProcess?.killed ?? true),
  );
  if (captainProcess && captainProcess.exitCode === null) {
    const success = killCaptain();
    if (success) {
      global.captainProcess = null;
      log.info("Successfully terminated captain :)");
    } else {
      log.error("Something went wrong when terminating captain!");
    }
  }
};

export const loadFileFromFullPath = async (
  filePath: string,
): Promise<string> => {
  return fs.readFileSync(filePath, { encoding: "utf-8" }).toString();
};

export const saveFileToFullPath = async (
  filePath: string,
  content: string,
): Promise<Result<void>> => {
  try {
    fs.writeFileSync(filePath, content);
    return Ok(undefined);
  } catch (error) {
    return Err(error as Error);
  }
};
