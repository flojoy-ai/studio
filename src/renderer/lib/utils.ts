import { type ClassValue, clsx } from "clsx";
import { twMerge } from "tailwind-merge";
import { z } from "zod";
import * as E from "fp-ts/Either";

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}

export function tryParse<Z extends z.ZodTypeAny>(z: Z) {
  return (val: unknown) => {
    const res = z.safeParse(val);
    if (res.success) {
      return E.right(res.data as z.infer<Z>);
    }
    return E.left(res.error);
  };
}
