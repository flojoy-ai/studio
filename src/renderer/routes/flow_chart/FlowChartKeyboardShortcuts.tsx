import useKeyboardShortcut from "@/renderer/hooks/useKeyboardShortcut";
import { useLoadApp } from "@/renderer/hooks/useLoadApp";
import { useSave } from "@/renderer/hooks/useSave";
import { useReactFlow } from "reactflow";

const FlowChartKeyboardShortcuts = () => {
  const { zoomIn, zoomOut } = useReactFlow();
  const save = useSave();
  const openFileSelector = useLoadApp();

  useKeyboardShortcut("ctrl", "=", zoomIn);
  useKeyboardShortcut("meta", "=", zoomIn);

  useKeyboardShortcut("ctrl", "-", zoomOut);
  useKeyboardShortcut("meta", "-", zoomOut);

  useKeyboardShortcut("ctrl", "s", save);
  useKeyboardShortcut("meta", "s", save);

  useKeyboardShortcut("ctrl", "o", openFileSelector);
  useKeyboardShortcut("meta", "o", openFileSelector);

  return <div></div>;
};

export default FlowChartKeyboardShortcuts;
