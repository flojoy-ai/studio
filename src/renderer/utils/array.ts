export function filterMap<T, U>(
  arr: Array<T>,
  mapPred: (x: T) => U | undefined | null,
) {
  return arr.reduce((acc: Array<U>, x: T) => {
    const val = mapPred(x);
    if (val !== undefined && val !== null) {
      acc.push(val);
    }

    return acc;
  }, []);
}
