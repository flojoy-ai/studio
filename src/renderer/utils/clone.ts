import { DeepMutable } from "@/renderer/types/util";

export const deepMutableClone = <T>(obj: T): DeepMutable<T> => {
  return JSON.parse(JSON.stringify(obj)) as DeepMutable<T>;
};
