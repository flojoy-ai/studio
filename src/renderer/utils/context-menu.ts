export const calculateContextMenuOffset = (
  clickX: number,
  clickY: number,
  flow: HTMLDivElement,
) => {
  // Calculate position of the context menu. We want to make sure it
  // doesn't get positioned off-screen.
  const pane = flow.getBoundingClientRect();

  const contextMenuHeight = 200;
  const paneToBlock = clickY - contextMenuHeight;

  let top: number | undefined = undefined;
  let bottom: number | undefined = undefined;

  if (paneToBlock < contextMenuHeight / 2) {
    top = paneToBlock;
  } else if (paneToBlock < contextMenuHeight) {
    if (pane.height - paneToBlock < contextMenuHeight) {
      top = contextMenuHeight - paneToBlock;
    } else {
      top = paneToBlock;
    }
  } else if (pane.height - paneToBlock < contextMenuHeight) {
    top = undefined;
    bottom = pane.height - paneToBlock;
  } else {
    top = paneToBlock;
  }

  return {
    top,
    left: clickX < pane.width - 200 ? clickX : undefined,
    right: clickX >= pane.width - 200 ? pane.width - clickX : undefined,
    bottom,
  };
};
