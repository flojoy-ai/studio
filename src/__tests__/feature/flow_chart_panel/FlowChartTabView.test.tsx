import { renderWithTheme } from "@src/__tests__/__utils__/utils";
import FlowChartKeyboardShortcuts from "@src/feature/flow_chart_panel/FlowChartKeyboardShortcuts";
import FlowChartTab from "@src/feature/flow_chart_panel/FlowChartTabView";

class ResizeObserver {
  observe() {}
  unobserve() {}
}

class IntersectionObserver {
  constructor() {}

  observe() {
    return null;
  }

  disconnect() {
    return null;
  }

  unobserve() {
    return null;
  }
}

jest.mock("@src/feature/flow_chart_panel/views/NodeModal", () => {
  const mockChildren = jest
    .fn()
    .mockReturnValue(<div data-testid="node-modal" />);
  return { __esModule: true, default: mockChildren };
});

jest.mock("@src/services/FlowChartServices", () => {
  return {
    saveFlowChartToLocalStorage: () => jest.fn(),
  };
});

jest.mock("@src/feature/flow_chart_panel/FlowChartKeyboardShortcuts", () => {
  return {
    default: jest.fn(() => {
      return <div></div>;
    }),
  };
});

jest.mock("@src/feature/flow_chart_panel/FlowChartTabState", () => {
  return {
    useFlowChartTabState: jest.fn().mockReturnValue({
      windowWidth: 1024,
      modalIsOpen: false,
      setIsModalOpen: jest.fn(),
      openModal: jest.fn(),
      afterOpenModal: jest.fn(),
      closeModal: jest.fn(),
      nd: {},
      setNd: jest.fn(),
      nodeLabel: "PYTHON FUNCTION",
      nodeType: "PYTHON FUNCTION TYPE",
      setNodeLabel: jest.fn(),
      setNodeType: jest.fn(),
      setPythonString: jest.fn(),
      setNodeFileName: jest.fn(),
      nodeFileName: "test.py",
      pythonString: "...",
      defaultPythonFnLabel: "PYTHON FUNCTION",
      defaultPythonFnType: "PYTHON FUNCTION TYPE",
      setNodeFilePath: jest.fn(),
    }),
  };
});

jest.mock("@src/feature/flow_chart_panel/FlowChartTabEffects", () => {
  return {
    useFlowChartTabEffects: () => {},
  };
});

jest.mock("@src/configs/NodeConfigs", () => {
  return {
    default: jest.fn(),
  };
});

jest.mock("@src/hooks/useFlowChartState");
jest.mock("@src/hooks/useControlsState");
jest.mock("@src/hooks/useSocket");

jest.mock("@src/utils/ManifestLoader", () => {
  return {
    CMND_TREE: { title: "ROOT", child: [] },
    CMND_MANIFEST_MAP: {},
    FUNCTION_PARAMETERS: {},
    getManifestCmdsMap: jest.fn(),
    getManifestParams : jest.fn()
  };
});

jest.mock("react-router-dom");

jest.mock("@src/data/pythonFunctions.json", () => ({
  __esModule: true,
  default: {},
}));

jest.mock("@src/configs/NodeConfigs", () => ({
  __esModule: true,
  nodeConfigs: {},
}));

window.ResizeObserver = ResizeObserver as any;
window.IntersectionObserver = IntersectionObserver as any;

describe("FlowChartTabView", () => {
  it("should render the component correcty", () => {
    const { container } = renderWithTheme(<FlowChartTab />);
    expect(container).toMatchSnapshot();
  });

  it.each([
    ["ReactFlow Provider", "react-flow-provider"],
    ["react-flow", "react-flow"],
    ["NodeModal component", "node-modal"],
  ])("should contain %p component", (msg, testId) => {
    const { getByTestId, getAllByTestId } = renderWithTheme(<FlowChartTab />);

    let component;

    if (testId === "react-flow") {
      component = getAllByTestId(testId);
      expect(component).toHaveLength(2);
    } else {
      component = getByTestId(testId);
      expect(component).toBeInTheDocument();
    }
  });

  it("checks the reactflow style", () => {
    const { getAllByTestId } = renderWithTheme(<FlowChartTab />);

    const component = getAllByTestId("react-flow")[0];
    expect(component).toHaveStyle("height: calc(100vh - 150px)");
  });
});
