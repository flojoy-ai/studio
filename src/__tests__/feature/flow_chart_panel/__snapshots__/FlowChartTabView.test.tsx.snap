// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowChartTabView should render the component correcty 1`] = `
<div>
  <div
    data-testid="react-flow-provider"
  >
    <div
      data-rfinstance="[]"
      data-testid="react-flow"
      style="height: 99vh;"
    >
      <div
        data-testid="react-flow"
      />
    </div>
    <div
      data-testid="node-modal"
    />
  </div>
</div>
`;
